anndata/tests/test_hdf5_backing.py

```diff
@@ -90,23 +90,23 @@ def test_read_write_X(tmp_path, mtx_format, backed_mode, force_dense):
 
 # this is very similar to the views test
 def test_backing(adata, tmp_path, backing_h5ad):
-    assert not adata.isbacked
+    assert not adata.filename
```

Comment:
In that case “filename” is wrong too, and we should go to `backing: Union[Path, UnknownFutureObjectStore, None]`.

Reply:
In principle I agree with that, though I'd take this further and say this should be expressed at a type level. Unfortunately I don't think we can actually do that (since you'd need to change the type of an object).

Just for ease of use, I like functions like `is_*` for convenience in the API. `if adata.is_backed` is a little nicer than `if adata.backed is not None` (since `if val` can throw an error).

> I’m allergic to mutually exclusive properties (like if x is None y needs to be set): That stuff is untestable. Rather there should be a single enum property that can take all possible states.

I'd probably go for `adata.store: Union[AnnDataFileManager, ...]`, then the filename could be accessed from there. That said, I'd like to have some time playing around with what we can do with backed mode before making strong choices about an API.

Reply:
`if adata.backing` or `if adata.store` is even shorter than `if adata.is_backed`.

`is not None` is only necessary if other falsy values should be treated differently than `None`.

Reply:
> if adata.backing or if adata.store is even shorter than if adata.is_backed

Yes, but they could be ambiguous. `is_backed` is like the `pd.api.type.is_*`, should be completely unambiguous.

----------------

> is not None is only necessary if other falsy values should be treated differently than None.

Which could happen.

To quote from the [google python style guide](http://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations):

> Always use if foo is None: (or is not None) to check for a None value-e.g., when testing whether a variable or argument that defaults to None was set to some other value. The other value might be a value that’s false in a boolean context!

--------------------

> Are there APIs in AnnDataFileManager that should be exposed? Else I’d go with store/backing: Union[Path, Any, None] (the Any for future extensibility)

Maybe? I think backed mode implementation needs to be considered, and I'd wait on API changes until that happened.